W-15059913-rtf-monitoring-kt (#1242)
mulesoft/docs-runtime-fabric: modules/ROOT/pages/use-anypoint-monitoring.adoc

For Mule apps deployed or managed using Runtime Fabric agent version 2.5.x or lower, two different agents are used as follows:

* One DaemonSet per underlying node to transport logs to Anypoint monitoring control plane.
* One sidecar per application replica to transporting metrics to Anypoint monitoring control plane. 

For optimal daemonset functionality, follow these prerequisites:

. Check that the application containers write logs to the following locations:

* `/var/log/pods`
* `/var/log/containers`
* `/var/lib/docker/containers`

. Check that the DaemonSet creates and writes on the host path `/var/lib/filebeat-data`.
. Ensure that the Docker daemon config includes the `json-file` logging driver.
. Ensure that the Anypoint Monitoring log forwarder DaemonSet container has write access to the host on the host path `/var/lib/filebeat-data`.

For applications deployed or managed using Runtime Fabric versions 2.6.x and above, MuleSoft consolidates telemetry transport to a single sidecar per application replica.

Dependent on the features you use, the sidecar agent consumes corresponding resources for optimal performance:

> 

Reviewer: Translate these two points into bullets:

- To enable default Anypoint metrics, the sidecar agent (per application replica) requires 0.05vCPU or Core of CPU allocation and 90MiB memory allocation.
- If you opt into xref:runtime-fabric/deploy-to-runtime-fabric.adoc#configure-logging-and-log-levels[forwarding application logs to Anypoint Monitoring], the sidecar agent (per application replica) requires up to 30% of CPU limit configured for application replica and up to 150MiB memory allocation

> To enable default Anypoint metrics, the sidecar agent (per application replica) requires 0.05vCPU or Core of CPU allocation and 90MB memory allocation.

Reviewer 2: 90MiB

> If you opt into xref:runtime-fabric/deploy-to-runtime-fabric.adoc#configure-logging-and-log-levels[forwarding application logs to Anypoint Monitoring], the sidecar agent (per application replica) requires up to 30% of CPU limit configured for application replica and up to 150MB memory allocation

Reviewer 2: this is 150MiB (not MB)

> Allocate capacity to the underlying cluster by using maximum configurations to avoid application deployment failure.

Reviewer: Note: Please allocate appropriate capacity to the underlying cluster to avoid application deployment failure.

= Using Anypoint Monitoring

> Mule apps and API gateways deployed to Runtime Fabric include native support for Anypoint Monitoring to allow viewing metrics within Anypoint Platform.

Reviewer 3: ```suggestion
Mule apps and API gateways deployed to Runtime Fabric include native support for Anypoint Monitoring that allows viewing metrics from within Anypoint Platform.
```

By default, Runtime Fabric is configured to send application metrics to Anypoint Monitoring. Data is securely transmitted using mutual TLS encryption. For metrics and logs to be sent to Anypoint Monitoring, outbound network connectivity must be available. 

If you need an outbound proxy, configure Anypoint Runtime Fabric to use a SOCKS5 proxy by following the instructions in xref:manage-proxy.adoc[Manage Proxies Used by Runtime Fabric].

To verify outbound network connectivity on Anypoint Runtime Fabric version 1.8.50 or later, ensure that each node in Runtime Fabric can send outbound TCP (Lumberjack) communication using port 443 to the following domains:

* US control plane: `dias-ingestor-router.us-east-1.prod.cloudhub.io` 
* EU control plane: `dias-ingestor-router.eu-central-1.prod-eu.msap.io`

[NOTE]
====
If you are using a version of Anypoint Runtime Fabric older than 1.8.50, use port 5044 instead of port 443. Use the following domains:

* US control plane: `dias-ingestor-nginx.prod.cloudhub.io` 
* EU control plane: `dias-ingestor-nginx.prod-eu.msap.io`
====

== Metrics

Each Mule app and API gateway replica runs a metrics agent that collects and transmits metrics data to 

> Anypoint Monitoring via secure mutual TLS encryption. Each replica can cache up to 100 megabytes of metrics, after which the metrics data is rotated to provide storage for more recent data. In the event that metrics cannot be sent to Anypoint Monitoring, this typically provides at least 4 days worth of storage per replica. Note that custom metrics might consume more storage.

Reviewer 3: ```suggestion
Anypoint Monitoring via secure mutual TLS encryption. Each replica can cache up to 100 megabytes of metrics, after which the metrics data is rotated to provide storage for more recent data. If metrics can't be sent to Anypoint Monitoring, this typically provides at least four days worth of storage per replica. Custom metrics might consume more storage.
```

> [NOTE]

Reviewer 3: I would have this as regular text instead of a note.

--

> The following Anypoint Runtime Fabric updates do not impact or restore monitoring metrics of applications deployed on Runtime Fabric:

Reviewer 3: ```suggestion
The following Anypoint Runtime Fabric updates do not impact or restore the monitoring metrics of applications deployed on Runtime Fabric:
```

Reviewer 3: What do you mean that the updates don't restore the monitoring metrics? If they don't impact the metrics, wouldn't they not restore them as well?

Author: @SparshAgarwal for confirmation

Author: Confirmed that metrics are owned by control plane. rtf backup restore is a runtime plane operation.

* A backup and restore of Runtime Fabric
* An upgrade of Runtime Fabric
* OS patching of Runtime Fabric nodes
--

== Logs 

Anypoint Monitoring provides access to xref:monitoring::logs.adoc[log data] for applications deployed to Rutime Fabric. To access logs with Anypoint Monitoring, you need a https://www.mulesoft.com/anypoint-pricing[Titanium subscription]. 

Each Mule app and API gateway replica can store up to 450 megabytes of log data on disk, after which log data 
is rotated to provide storage for more recent log data.

Logs generated by each replica of a Mule app and API gateway are collected and written to disk on the worker 

> node on which they are running. Each worker node runs a single log forwarding agent, which monitors this location 

Reviewer 3: What location is "this location" referring to?

and transmits the log data to Anypoint Monitoring.

[NOTE]
====

> The log forwarding agent can transmit up to 300 kilobytes of log data per second. You should not deploy more than

Reviewer 3: I would have this paragraph be regular text.

50 replicas per worker node.

In some cases, logs with multiple lines might not be fully transmitted to Anypoint Monitoring. This can occur if 
multiple lines of log output are split across two files due to log rotation inside Runtime Fabric.
====

You can enable log forwarding and set log levels when you xref:deploy-to-runtime-fabric.adoc#configure-logging-and-log-levels[deploy applications in Runtime Manager]. 

=== Resource and Permission Requirements for Anypoint Monitoring Agent

> MuleSoft runs Anypoint Monitoring agents alongside Mule apps to send telemetry to Anypoint control plane.

Reviewer 3: ```suggestion
MuleSoft runs Anypoint Monitoring agents alongside Mule apps to send telemetry to the Anypoint control plane.
```

For Mule apps deployed or managed using Runtime Fabric agent version 2.5.x or lower, two different agents are used as follows:

> * One daemonset per underlying node to transporting logs to Anypoint monitoring control plane.

Reviewer 3: ```suggestion
* One daemonset per underlying node to transport logs to Anypoint monitoring control plane
```

Reviewer 3: I looked up daemonset because I hadn't heard of it and I saw that it is usually written like this: DaemonSet.

> * One sidecar per application replica to transporting metrics to Anypoint monitoring control plane. 

Reviewer 3: ```suggestion
* One sidecar per application replica to transport metrics to Anypoint monitoring control plane
```

For optimal daemonset functionality, follow these prerequisites:

. Check that the application containers write logs to the following locations:

* `/var/log/pods`
* `/var/log/containers`
* `/var/lib/docker/containers`

> . Check the daemon-set creates and writes on the host path `/var/lib/filebeat-data`.

Reviewer 3: ```suggestion
. Check that the DaemonSet creates and writes on the host path `/var/lib/filebeat-data`.
```

. Ensure that the Docker daemon config includes the `json-file` logging driver.

> . Ensure that the Anypoint Monitoring log forwarder daemon-set container has write access to the host on the host path `/var/lib/filebeat-data`.

Reviewer 3: ```suggestion
. Ensure that the Anypoint Monitoring log forwarder DaemonSet container has write access to the host on the host path `/var/lib/filebeat-data`.
```

> For applications deployed or managed using Runtime Fabric versions 2.6.x and above, MuleSoft is unifying telemetry transport to a single sidecar per application replica.

Reviewer 3: ```suggestion
For applications deployed or managed using Runtime Fabric versions 2.6.x and above, MuleSoft consolidates telemetry transport to a single sidecar per application replica.
```

> * To enable default Anypoint metrics, the sidecar agent (per application replica) requires 0.05vCPU or Core of CPU allocation and 90MiB memory allocation.

Reviewer 4: @SparshAgarwal can you proof read and confirm once?

> To accommodate added functionality, this sidecar now requires higher resource allocation, ranging from 0.1CPU (minimum) to 30% of CPU Limit configured for application replica (maximum) to accommodate for needs of the Mule app. Memory consumption may range from 70 MB (minimum) to 150 MB (maximum).

Reviewer 3: ```suggestion
To accommodate added functionality, this sidecar requires higher resource allocation, ranging from 0.1CPU (minimum) to 30% of CPU Limit configured for application replica (maximum), depending on the needs of the Mule app. Memory consumption can range from 70 MB (minimum) to 150 MB (maximum).
```

> To accommodate added functionality, this sidecar now requires higher resource allocation, ranging from 0.1CPU (minimum) to 30% of CPU Limit configured for application replica (maximum) to accommodate for needs of the Mule app. Memory consumption may range from 70 MB (minimum) to 150 MB (maximum).

Reviewer: @kevintroller can we change "To accommodate added functionality, this sidecar now requires higher resource allocation, ranging from 0.1CPU (minimum) to 30% of CPU Limit configured for application replica (maximum) to accommodate for needs of the Mule app. Memory consumption may range from 70 MB (minimum) to 150 MB (maximum)." statement to below:
Dependent on the features used by customers, the sidecar agent consumes corresponding resources for optimal performance:
-  To enable default Anypoint metrics, the sidecar agent (per application replica) requires 0.05vCPU/Core of CPU allocation and 90MB memory allocation
- If customer opts into forwarding application logs to Anypoint Monitoring<link: https://docs.mulesoft.com/runtime-fabric/latest/deploy-to-runtime-fabric#configure-logging-and-log-levels>, the sidecar agent (per application replica) requires upto 30% of CPU Limit configured for application replica and upto 150MB memory allocation

> MuleSoft recommends to allocate appropriate capacity to the underlying cluster calculated using maximum configurations to avoid application deployment failure.

Reviewer 3: ```suggestion
Allocate capacity to the underlying cluster by using maximum configurations to avoid application deployment failure.
```

== Anypoint Visualizer

xref:visualizer::index.adoc[Anypoint Visualizer] displays views of different aspects of an application network graph. 
You can use the graph to explore your application network, identify problems, and make decisions.

[WARNING]
====
By default, Anypoint Visualizer is disabled when deploying an application to Mule runtime 4.x. 
See xref:visualizer::setup.adoc#enable-header-injection[Enable Header Injection] for more information.
Due to the impact on performance, enabling Anypoint Visualizer for application deployments with less than 0.2 vCPU is 
not supported.
====

== (Optional) Disable Anypoint Monitoring for a Mule App

When you deploy a Mule app to Runtime Fabric, Anypoint Monitoring is enabled by default. To disable Anypoint 
Monitoring, pass the following custom property when deploying your Mule app:

[source,copy]
----
anypoint.platform.config.analytics.agent.enabled=false
----

[NOTE]

> For compliance or geospecific requirement if your team or organization cannot send metrics to Anypoint control plane. You can disable Anypoint metrics if you are not contractually obligated to send these metrics for Anypoint usage and billing purposes.

Reviewer 3: It is not clear what you are saying in the first sentence.

Author: @SparshAgarwal can you reconfirm the first sentence meaning?
